packages/fx-core/src/core/middleware/projectUpgrader.ts

```diff
@@ -0,0 +1,202 @@
+      );
+    }
+
+    // Read UserData file.
+    const userDataPath = path.resolve(confFolderPath, `${envName}.userdata`);
+    const userData = await readUserData(userDataPath);
+
+    // Merge updatedKeys into UserData.
+    mergeKeysToUserDate(userData, updatedKeys);
+
+    // Save UserData
+    await saveUserData(userDataPath, userData);
+  } catch (error) {
+    ctx.result = err(ContextUpgradeError(error));
+  }
+}
+
+async function readUserData(userDataPath: string): Promise<Record<string, string>> {
+  let dict: Record<string, string>;
+  if (await fs.pathExists(userDataPath)) {
+    const dictContent = await fs.readFile(userDataPath, "UTF-8");
+    dict = deserializeDict(dictContent);
+  } else {
+    dict = {};
+  }
+
+  return dict;
+}
+
+async function saveUserData(userDataPath: string, userData: Record<string, string>): Promise<void> {
+  await fs.writeFile(userDataPath, serializeDict(userData));
+}
+
+async function readContext(contextPath: string): Promise<Json> {
+  const configJson: Json = await fs.readJson(contextPath);
+  return configJson;
+}
+
+async function saveContext(contextPath: string, context: Json): Promise<void> {
+  await fs.writeFile(contextPath, JSON.stringify(context, null, 4));
+}
+
+function updateContextValue(context: Json): Map<string, any> {
+  const res: Map<string, any> = new Map();
+
+  // Update resource context.
+  for (const item of resourceContext) {
+    const pluginContext: any = context[item.plugin];
+    if (!pluginContext) {
+      continue;
+    }
+
+    for (const key of item.relatedKeys) {
+      // Save value to res and update value to secret pattern if value is not in secret pattern.
+      if (pluginContext[key] && !isSecretPattern(pluginContext[key])) {
+        res.set(getUserDataKey(item.plugin, key), pluginContext[key]);
+        pluginContext[key] = getSecretPattern(item.plugin, key);
+      }
+    }
+  }
+
+  // Update solution context.
+  const pluginContext: any = context[solutionContext.plugin];
+  for (const key of solutionContext.relatedKeys) {
+    if (pluginContext[key] && !isSecretPattern(pluginContext[key])) {
+      res.set(getUserDataKey(solutionContext.plugin, key), pluginContext[key]);
+      pluginContext[key] = getSecretPattern(solutionContext.plugin, key);
+    }
+  }
+
+  return res;
+}
+
+function mergeKeysToUserDate(
+  userData: Record<string, string>,
+  updatedKeys: Map<string, any>
+): void {
+  // Move resource context first to userdata
+  let moved = false;
+  for (const item of resourceContext) {
+    // Check whether corresponding secret exists.
+    if (!userData[getUserDataKey(item.plugin, item.secret)]) {
+      continue;
+    }
+
+    for (const key of item.relatedKeys) {
+      const userDataKey = getUserDataKey(item.plugin, key);
+      // Merge will only happen when userData does not contain certain key.
+      // Otherwise, value in userData will be regarded as source of truth.
```

Comment:
Though it's very rare case, but this may change local debug behavior. Is PM aware of this?

Reply:
This scenario means that this user has already run local debug with latest extension, and is using a env.default.json that created with old extension (e.g. pulled code from github and the other user is using old extension). We will treat info in userdata as source of truth since this file is generated by this user.